app/master/cluster_master.py

```diff
@@ -143,6 +144,15 @@ def handle_slave_state_update(self, slave, new_slave_state):
         do_transition = slave_transition_functions.get(new_slave_state)
         do_transition(slave)
 
+    def _shutdown_slave(self, slave):
+        """
+        Puts slave in shutdown mode so it cannot receive new builds. The slave will be killed when finished with any
+        running builds.
+        :type slave: Slave
+        """
+        slave.set_shutdown_mode()
+        self._logger.info('Slave on {} was put in shutdown mode. (id: {})', slave.url, slave.id)
```

Comment:
I'm fine with punting on this for now. You're right -- it sounds trickier than I thought originally and is really just a nice to have that we can add later.